lib/sequelize.js

```diff
@@ -974,41 +975,61 @@ class Sequelize {
     // testhint argsConform.end
 
     const transaction = new Transaction(this, options);
-    const ns = Sequelize.cls;
-
-    if (autoCallback) {
-      let transactionResolver = (resolve, reject) => {
-        transaction.prepareEnvironment().then(() => {
-          if (ns) {
-            autoCallback = ns.bind(autoCallback);
-          }
-
-          const returnValue = autoCallback(transaction);
-          if (!returnValue || !returnValue.then) throw new Error('You need to return a promise chain/thenable to the sequelize.transaction() callback');
-
-          return returnValue.then(result => transaction.commit()).then(() => {
-            resolve(returnValue);
-          });
-        }).catch(err => {
-          // If the transaction has already finished (commit, rollback, etc), reject with the original error
-          if (transaction.finished) {
-            reject(err);
-          } else {
-            return transaction.rollback().finally(() => {
-              reject(err);
-            });
-          }
-        });
-      };
-
-      if (ns) {
-        transactionResolver = ns.bind(transactionResolver, ns.createContext());
-      }
+    const ns = Sequelize._cls;
+
+    if (!autoCallback) return transaction.prepareEnvironment().return(transaction);
+
+    // autoCallback provided
+    const wrapper = ns ? function(fn) {
+      var promise;
+      ns.run(() => promise = fn());
+      return promise;
+    } : function(fn) { return fn(); };
```

Comment:
Could the wrapper be defined outside transaction and simply check for Sequelize._cls? We'll save a function allocation for each transaction.